src/local-context.ts

```diff
@@ -0,0 +1,14 @@
+import * as vscode from 'vscode';
+
+/**
+ *  A class that lets us keep track of disposables, independently from the
+ *  extension's global context.
+ */
+export class LocalContext implements vscode.Disposable {
```

Comment:
Thinking more about this, I feel like we should probably move `clangdClient` into the `Context`, 

What do you think the proposed API below?

```
export class ClangdContext implements vscode.Disposable {
   subscriptions: vscode.Disposable[] = [];
   client: ClangdLanguageClients
   globalStoragePath: string

   activate() {
      // do whatever `startClient` does, get client & server options, init client, call client.start etc.
   }

   dispose() {
      ...
   }
}
```

This seems to simplify the codebase:
- the `client` parameter in `activate` is not needed anymore,  it can be fetched via `clangdContext` object (the same for the `globalStoragePath`);
- the `startDisposable` in `clangdLanuageClient` can be removed;
- the restart command implement logic seems more natural; 

But it would require more refactoring changes, I think it is fine.

Reply:
Thanks, that's much nicer.

I pushed a new version of the branch (it's totally rewritten, this couldn't be done with fixup commits in any useful way, sorry for that). I decided to leave `globalStoragePath` the way it is, since it's only used in constructors, so I thought it's not useful to store it in the class.

The first commit is a bit big and the diff is hard to read because of the large amount of moved code. Let me know if you have any suggestions for improving that.